File: web/src/domain/openOptions/placeholders.ts
const PLACEHOLDER = /\[([+-]?)(\d+)(?:\.(\d+))?([%％])?\]/g;

export type Placeholder = {
  sign: string;
  index: number;
  precision: number;
  suffix: string;
};

export function placeholders(template: string): Placeholder[] {
  return [...template.matchAll(PLACEHOLDER)].map((match) => ({
    sign: match[1],
    index: Number(match[2]),
    precision: Number(match[3] || 0),
    suffix: match[4] || "",
  }));
}

export function titleTemplate(template: string, bindings: Record<string, number> = {}): string {
  const sourceIndex = (index: number) => bindings[String(index)] ?? index;
  const uniqueIndices = [...new Set(placeholders(template).map((placeholder) => sourceIndex(placeholder.index)))];
  const names = new Map(uniqueIndices.map((index, position) => [index, uniqueIndices.length === 1 ? "n" : `n${position + 1}`]));
  return template.replace(PLACEHOLDER, (_, sign: string, index: string, _precision: string, suffix: string) => `[${sign}${names.get(sourceIndex(Number(index)))}${suffix || ""}]`);
}

export function hasPlaceholder(template: string): boolean {
  PLACEHOLDER.lastIndex = 0;
  return PLACEHOLDER.test(template);
}

export { PLACEHOLDER };

File: src/rs_dev/open_options/templates/placeholders.py
"""Parse display placeholders without depending on a locale."""

from __future__ import annotations

import re


PLACEHOLDER = re.compile(r"\[([+-]?)(\d+)(?:\.(\d+))?([%％])?\]")


def placeholder_indices(template: str) -> list[int]:
    return [int(match.group(2)) for match in PLACEHOLDER.finditer(template)]


def title_template(template: str) -> str:
    unique_indices = list(dict.fromkeys(placeholder_indices(template)))
    names = {
        index: "n" if len(unique_indices) == 1 else f"n{position + 1}"
        for position, index in enumerate(unique_indices)
    }

    def replace(match: re.Match[str]) -> str:
        sign, raw_index, _precision, suffix = match.groups()
        return f"[{sign}{names[int(raw_index)]}{suffix or ''}]"

    return PLACEHOLDER.sub(replace, template)
